dial.go

```diff
@@ -0,0 +1,148 @@
+	KeepAlive time.Duration
+
+	// DNS server(s) to use instead of system default.
+	// If provided with multiple servers, the first one is used as primary server, the rest are used as a fallback.
+	DNSServers []netip.AddrPort
+
+	// Timeout is the timeout for DNS queries. Works only with user-provided DNS servers.
+	// Each server is given ResolveTimeout to respond.
+	ResolveTimeout time.Duration
+}
+
+func DefaultDialConfig() DialConfig {
+	return DialConfig{
+		DialTimeout:    10 * time.Second,
+		KeepAlive:      30 * time.Second,
+		ResolveTimeout: 5 * time.Second,
+	}
+}
+
+func (c *DialConfig) Validate() error {
+	for i, u := range c.DNSServers {
+		if err := validateDNSAddress(u); err != nil {
+			return fmt.Errorf("servers[%d]: %w", i, err)
+		}
+	}
+	return nil
+}
+
+type dialFunc func(context.Context, string, string) (net.Conn, error)
+
+type Dialer struct {
+	cfg       DialConfig
+	nd        *net.Dialer
+	resolvers []net.Resolver
+	log       log.Logger
+}
+
+func NewDialer(cfg DialConfig, l log.Logger) (*Dialer, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	if len(cfg.DNSServers) > 0 {
+		l.Infof("using DNS servers: %s", cfg.DNSServers)
+	}
+
+	nd := &net.Dialer{
+		Timeout:   cfg.DialTimeout,
+		KeepAlive: cfg.KeepAlive,
+	}
+
+	resolvers := make([]net.Resolver, 0, len(cfg.DNSServers))
+	for _, s := range cfg.DNSServers {
+		dnsAddress := s.String()
+		resolvers = append(resolvers, net.Resolver{
+			PreferGo: true,
+			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+				l.Debugf("dial DNS server %s instead of %s", dnsAddress, address)
+				return nd.DialContext(ctx, network, dnsAddress)
+			},
+		})
+	}
+
+	return &Dialer{
+		cfg:       cfg,
+		nd:        nd,
+		resolvers: resolvers,
+		log:       l,
+	}, nil
+}
+
+func (d *Dialer) LookupIP(ctx context.Context, network, hostname string) ([]net.IP, error) {
+	for i := range d.resolvers {
+		lctx, cancel := context.WithTimeout(ctx, d.cfg.ResolveTimeout)
+		ips, err := d.resolvers[i].LookupIP(lctx, network, hostname)
+		cancel()
+		if err != nil {
+			d.log.Debugf("failed to resolve %s on %s: %s", hostname, d.cfg.DNSServers[i], err)
+		} else {
+			return ips, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to resolve %s on any DNS server", hostname)
+}
+
+func (d *Dialer) DialWithLookup(ctx context.Context, network, address string) (net.Conn, error) {
+	host, port, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+
+	ips, err := d.LookupIP(ctx, "ip", host)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ip := range ips {
+		address := ip.String() + ":" + port
+		conn, err := d.nd.DialContext(ctx, network, address)
+		if err != nil {
+			d.log.Debugf("failed to dial %s: %s", address, err)
+		} else {
+			return conn, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to dial %s on any IP address", address)
+}
+
+func (d *Dialer) DialFunc() dialFunc { //nolint:revive,golint // Unexported by design.
```

Comment:
I agree with the linter. It looks odd and has no purpose really. The Dialer should implement DialContext.